---
title: VCF BCF Format
date: 2024-04-16
description: ""
tags:
  - genomics
  - ngs
---
[VCF document](https://samtools.github.io/hts-specs/VCFv4.2.pdf)

VCF - Variant Call Format, is a text-based, tab-delimited format commonly used for variant calling.
BCF is the binary version of VCF.
VCF uses 1-based indices while BCF uses 0-based indices.

VCF files can be generated by tools such as samtools, GATK, FreeBayes, and can be analyzed/visualized by tools such as vcftools, bcfools and ANNOVAR.

Meta-information lines: "##" at the beginning of a file 
  1. fileformat - VCF version, required first line
  2. INFO - 4 required tags: ID, Number, Type, Description
  3. FILTER 
  4. FORMAT
  5. ALT - alternative allele format, ID can be DEL, INS, DUP, INV, DUP:TANDEM; DEL:ME, INS:ME
  6. assembly
  7. contig - a contiguous sequence, usually a reference genome/chromosome assembly.
  8. SAMPLE
  9. PEDIGREE - relationships between genome or database url

Header line: names of 8 fixed mandatory columns.

Col numbers: 1 CHROM; 2 POS;  3 ID; 4 REF; 5 ALT; 6 QUAL; 7 FILTER; 8 INFO

VCF file example [from international genome](https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/)

```yaml
##fileformat=VCFv4.0
##fileDate=20090805
##source=myImputationProgramV3.1
##reference=1000GenomesPilot-NCBI36
##phasing=partial
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
##INFO=<ID=AF,Number=.,Type=Float,Description="Allele Frequency">
##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral Allele">
##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129">
##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">
##FILTER=<ID=q10,Description="Quality below 10">
##FILTER=<ID=s50,Description="Less than 50% of samples have data">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
#CHROM POS     ID        REF ALT    QUAL FILTER INFO                              FORMAT      NA00001        NA00002        NA00003
20     14370   rs6054257 G      A       29   PASS   NS=3;DP=14;AF=0.5;DB;H2           GT:GQ:DP:HQ 0|0:48:1:51,51 1|0:48:8:51,51 1/1:43:5:.,.
20     17330   .         T      A       3    q10    NS=3;DP=11;AF=0.017               GT:GQ:DP:HQ 0|0:49:3:58,50 0|1:3:5:65,3   0/0:41:3
20     1110696 rs6040355 A      G,T     67   PASS   NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ 1|2:21:6:23,27 2|1:2:0:18,2   2/2:35:4
```
